File: python3/same_tree.py
from .model import TreeNode

"""
Space   : O(n)
Time    : O(n)
"""



class Solution:
    def preorder(self, root: TreeNode) -> List[int]:
        s = []
        if root:
            s.append(root.val)
            s += self.preorder(root.left)
            s += self.preorder(root.right)
        else:
            return [None]
        return s

    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        lp = self.preorder(p)
        lq = self.preorder(q)

        return (lp == lq) and (len(lp) == len(lq))

File: golang/same_tree.go
package main

import "golang/model"

/*
Space	: O(n)
Time	: O(n)
*/

func isSameTree(p *model.TreeNode, q *model.TreeNode) bool {
	ap := []*model.TreeNode{p}
	aq := []*model.TreeNode{q}

	for len(ap) > 0 && len(aq) > 0 {
		np := ap[0]
		ap = ap[1:]
		nq := aq[0]
		aq = aq[1:]

		if np == nil && nq == nil {
			continue
		}
		if np != nil && nq != nil {
			if np.Val != nq.Val {
				return false
			}

			ap = append(ap, np.Left)
			ap = append(ap, np.Right)
			aq = append(aq, nq.Left)
			aq = append(aq, nq.Right)
		} else {
			return false
		}
	}

	return true
}
